File: frontend/src/api/worlds.ts
import type { components } from "./generated";
import { ApiError } from "./health";

export type GameViewResponse = components["schemas"]["GameViewResponse"];
export type LocationRuleCreate = components["schemas"]["LocationRuleCreate"];
export type LocationRuleResponse =
  components["schemas"]["LocationRuleResponse"];
type RoleResponse = components["schemas"]["RoleResponse"];
export type WorldCreate = components["schemas"]["WorldCreate"];
export type WorldResponse = components["schemas"]["WorldResponse"];
export type WorldRoleResponse = components["schemas"]["WorldRoleResponse"];

export const locationIds = [
  "the_home",
  "the_dungeon",
  "the_mall",
  "the_guild",
  "the_hotel",
  "the_school",
] as const;
export type LocationId = (typeof locationIds)[number];
export type SceneId = "the_world_map" | LocationId;

export interface WorldsApi {
  listWorlds(signal?: AbortSignal): Promise<WorldResponse[]>;
  createWorld(payload: WorldCreate): Promise<WorldResponse>;
  deleteWorld(worldId: number): Promise<void>;
  listWorldRoles(
    worldId: number,
    signal?: AbortSignal,
  ): Promise<WorldRoleResponse[]>;
  addNpc(worldId: number, roleId: number): Promise<WorldRoleResponse>;
  setNpcEnabled(
    worldId: number,
    roleId: number,
    enabled: boolean,
  ): Promise<WorldRoleResponse>;
  removeNpc(worldId: number, roleId: number): Promise<void>;
  replaceLocationRules(
    worldId: number,
    roleId: number,
    rules: LocationRuleCreate[],
  ): Promise<LocationRuleResponse[]>;
  selectLocation(
    worldId: number,
    locationId: LocationId,
  ): Promise<GameViewResponse>;
  getGameView(
    worldId: number,
    sceneId: SceneId,
    signal?: AbortSignal,
  ): Promise<GameViewResponse>;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isNumber(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value);
}

function isInteger(value: unknown): value is number {
  return isNumber(value) && Number.isInteger(value);
}

function isPositiveInteger(value: unknown): value is number {
  return isInteger(value) && value > 0;
}

function isNonNegativeInteger(value: unknown): value is number {
  return isInteger(value) && value >= 0;
}

function isScalar(value: unknown): value is number | string | boolean {
  return (
    isNumber(value) || typeof value === "string" || typeof value === "boolean"
  );
}

function isScalarRecord(
  value: unknown,
): value is Record<string, number | string | boolean> {
  return isRecord(value) && Object.values(value).every(isScalar);
}

function isLocationId(value: unknown): value is LocationId {
  return (
    typeof value === "string" &&
    locationIds.some((locationId) => locationId === value)
  );
}

function isSceneId(value: unknown): value is SceneId {
  return value === "the_world_map" || isLocationId(value);
}

function isTimeSlot(value: unknown): value is LocationRuleCreate["time_slot"] {
  return (
    value === "morning" ||
    value === "midday" ||
    value === "evening" ||
    value === "night"
  );
}

function isWeekday(value: unknown): boolean {
  return (
    value === "monday" ||
    value === "tuesday" ||
    value === "wednesday" ||
    value === "thursday" ||
    value === "friday" ||
    value === "saturday" ||
    value === "sunday"
  );
}

function isAttributeDefinition(value: unknown): boolean {
  if (!isRecord(value)) return false;
  const dataType = value.data_type;
  const operations = value.allowed_operations;
  return (
    typeof value.key === "string" &&
    typeof value.display_name === "string" &&
    (dataType === "integer" ||
      dataType === "number" ||
      dataType === "string" ||
      dataType === "boolean" ||
      dataType === "enum") &&
    isScalar(value.base_value) &&
    typeof value.description === "string" &&
    typeof value.update_rule === "string" &&
    Array.isArray(operations) &&
    operations.every(
      (operation) =>
        operation === "replace" ||
        operation === "increment" ||
        operation === "decrement",
    ) &&
    (value.minimum === null ||
      value.minimum === undefined ||
      isNumber(value.minimum)) &&
    (value.maximum === null ||
      value.maximum === undefined ||
      isNumber(value.maximum)) &&
    Array.isArray(value.enum_options) &&
    value.enum_options.every((option) => typeof option === "string") &&
    (value.update_example === null ||
      value.update_example === undefined ||
      typeof value.update_example === "string") &&
    (value.no_update_example === null ||
      value.no_update_example === undefined ||
      typeof value.no_update_example === "string")
  );
}

function isRoleResponse(value: unknown): value is RoleResponse {
  return (
    isRecord(value) &&
    isPositiveInteger(value.id) &&
    typeof value.name === "string" &&
    typeof value.persona === "string" &&
    typeof value.system_prompt === "string" &&
    typeof value.world_book === "string" &&
    Array.isArray(value.attributes) &&
    value.attributes.every(isAttributeDefinition) &&
    isScalarRecord(value.effective_base_values) &&
    typeof value.portrait_url === "string" &&
    Array.isArray(value.referenced_world_ids) &&
    value.referenced_world_ids.every(isPositiveInteger) &&
    isPositiveInteger(value.version) &&
    typeof value.created_at === "string" &&
    typeof value.updated_at === "string"
  );
}

function isWorldResponse(value: unknown): value is WorldResponse {
  return (
    isRecord(value) &&
    isPositiveInteger(value.id) &&
    typeof value.display_name === "string" &&
    value.display_name === `世界 ${value.id}` &&
    isPositiveInteger(value.active_branch_id) &&
    isPositiveInteger(value.day) &&
    isWeekday(value.weekday) &&
    isTimeSlot(value.time_slot) &&
    isRoleResponse(value.player_role) &&
    value.player_role.referenced_world_ids.includes(value.id) &&
    isNonNegativeInteger(value.npc_count) &&
    typeof value.last_played_at === "string"
  );
}

function isWorldRoleResponse(value: unknown): value is WorldRoleResponse {
  return (
    isRecord(value) &&
    isPositiveInteger(value.world_id) &&
    isPositiveInteger(value.role_id) &&
    typeof value.name === "string" &&
    (value.kind === "player" || value.kind === "npc") &&
    typeof value.enabled === "boolean" &&
    isScalarRecord(value.effective_attributes) &&
    typeof value.portrait_url === "string" &&
    isPositiveInteger(value.version) &&
    typeof value.updated_at === "string"
  );
}

function isLocationCandidate(value: unknown): boolean {
  return (
    isRecord(value) &&
    isLocationId(value.location_id) &&
    isPositiveInteger(value.weight)
  );
}

function isLocationRuleResponse(value: unknown): value is LocationRuleResponse {
  return (
    isRecord(value) &&
    isPositiveInteger(value.id) &&
    isPositiveInteger(value.world_id) &&
    isPositiveInteger(value.role_id) &&
    isInteger(value.weekday_mask) &&
    value.weekday_mask >= 1 &&
    value.weekday_mask <= 127 &&
    isTimeSlot(value.time_slot) &&
    (value.mode === "fixed" || value.mode === "random") &&
    isInteger(value.priority) &&
    typeof value.enabled === "boolean" &&
    Array.isArray(value.candidates) &&
    value.candidates.length > 0 &&
    (value.mode === "random" || value.candidates.length === 1) &&
    value.candidates.every(isLocationCandidate)
  );
}

function isMapLocation(value: unknown): boolean {
  return (
    isRecord(value) &&
    isLocationId(value.scene_id) &&
    typeof value.display_name === "string" &&
    isNonNegativeInteger(value.order) &&
    isNumber(value.anchor_x) &&
    value.anchor_x >= 0 &&
    value.anchor_x <= 1 &&
    isNumber(value.anchor_y) &&
    value.anchor_y >= 0 &&
    value.anchor_y <= 1
  );
}

function isGameViewResponse(value: unknown): value is GameViewResponse {
  return (
    isRecord(value) &&
    isPositiveInteger(value.world_id) &&
    isSceneId(value.scene_id) &&
    typeof value.background_url === "string" &&
    typeof value.fallback_background_url === "string" &&
    typeof value.player_marker_url === "string" &&
    isLocationId(value.player_location_id) &&
    Array.isArray(value.locations) &&
    value.locations.every(isMapLocation) &&
    Array.isArray(value.visible_roles) &&
    value.visible_roles.every(isWorldRoleResponse)
  );
}

function isWorldRoleFor(
  value: unknown,
  worldId: number,
  roleId?: number,
  kind?: WorldRoleResponse["kind"],
  enabled?: boolean,
): value is WorldRoleResponse {
  return (
    isWorldRoleResponse(value) &&
    value.world_id === worldId &&
    (roleId === undefined || value.role_id === roleId) &&
    (kind === undefined || value.kind === kind) &&
    (enabled === undefined || value.enabled === enabled)
  );
}

function isWorldRoleListFor(
  value: unknown,
  worldId: number,
): value is WorldRoleResponse[] {
  if (
    !Array.isArray(value) ||
    !value.every((role) => isWorldRoleFor(role, worldId))
  ) {
    return false;
  }
  const roleIds = new Set(value.map((role) => role.role_id));
  return (
    roleIds.size === value.length &&
    value.filter((role) => role.kind === "player").length === 1
  );
}

function isRuleListFor(
  value: unknown,
  worldId: number,
  roleId: number,
  expectedLength: number,
): value is LocationRuleResponse[] {
  if (
    !Array.isArray(value) ||
    value.length !== expectedLength ||
    !value.every(
      (rule) =>
        isLocationRuleResponse(rule) &&
        rule.world_id === worldId &&
        rule.role_id === roleId,
    )
  ) {
    return false;
  }
  return new Set(value.map((rule) => rule.id)).size === value.length;
}

function isGameViewFor(
  value: unknown,
  worldId: number,
  expectedSceneId?: SceneId,
  expectedPlayerLocationId?: LocationId,
): value is GameViewResponse {
  if (
    !isGameViewResponse(value) ||
    value.world_id !== worldId ||
    (expectedSceneId !== undefined && value.scene_id !== expectedSceneId) ||
    (expectedPlayerLocationId !== undefined &&
      value.player_location_id !== expectedPlayerLocationId) ||
    !value.visible_roles.every((role) => role.world_id === worldId)
  ) {
    return false;
  }
  const visibleRoleIds = new Set(
    value.visible_roles.map((role) => role.role_id),
  );
  const locationSceneIds = new Set(
    value.locations.map((location) => location.scene_id),
  );
  return (
    visibleRoleIds.size === value.visible_roles.length &&
    locationSceneIds.size === value.locations.length
  );
}

async function requestJson<T>(
  url: string,
  init: RequestInit,
  validate: (value: unknown) => value is T,
): Promise<T> {
  const response = await fetch(url, init);
  if (!response.ok) {
    throw await responseError(response);
  }
  let payload: unknown;
  try {
    payload = await response.json();
  } catch (error) {
    // 仅转换响应 JSON 语法错误；取消和网络错误保留原始语义。
    if (error instanceof SyntaxError) {
      throw new ApiError("世界响应格式无效。", response.status);
    }
    throw error;
  }
  if (!validate(payload)) {
    throw new ApiError("世界响应格式无效。", response.status);
  }
  return payload;
}

async function requestEmpty(url: string, init: RequestInit): Promise<void> {
  const response = await fetch(url, init);
  if (!response.ok) {
    throw await responseError(response);
  }
}

function safeErrorDetail(value: unknown): string | null {
  if (!isRecord(value)) return null;
  const detail = value.detail;
  const message =
    typeof detail === "string"
      ? detail
      : isRecord(detail) && typeof detail.message === "string"
        ? detail.message
        : null;
  if (!message) return null;
  const normalized = message.trim().replace(/[\r\n\t]+/gu, " ");
  return normalized.length > 0 ? normalized.slice(0, 500) : null;
}

async function responseError(response: Response): Promise<ApiError> {
  let detail: string | null = null;
  try {
    const payload: unknown = await response.json();
    detail = safeErrorDetail(payload);
  } catch {
    // 错误体不可解析时只保留稳定的通用消息和 HTTP 状态。
  }
  return new ApiError(detail ?? "世界请求失败。", response.status);
}

const jsonHeaders = {
  Accept: "application/json",
  "Content-Type": "application/json",
} as const;

export const worldsApi: WorldsApi = {
  async listWorlds(signal) {
    return requestJson(
      "/api/worlds",
      { headers: { Accept: "application/json" }, signal },
      (value): value is WorldResponse[] => {
        if (!Array.isArray(value) || !value.every(isWorldResponse))
          return false;
        return new Set(value.map((world) => world.id)).size === value.length;
      },
    );
  },
  async createWorld(payload) {
    return requestJson(
      "/api/worlds",
      { method: "POST", headers: jsonHeaders, body: JSON.stringify(payload) },
      (value): value is WorldResponse =>
        isWorldResponse(value) &&
        value.day === 1 &&
        value.weekday === "monday" &&
        value.time_slot === "morning",
    );
  },
  async deleteWorld(worldId) {
    return requestEmpty(`/api/worlds/${worldId}`, { method: "DELETE" });
  },
  async listWorldRoles(worldId, signal) {
    return requestJson(
      `/api/worlds/${worldId}/roles`,
      { headers: { Accept: "application/json" }, signal },
      (value): value is WorldRoleResponse[] =>
        isWorldRoleListFor(value, worldId),
    );
  },
  async addNpc(worldId, roleId) {
    return requestJson(
      `/api/worlds/${worldId}/roles`,
      {
        method: "POST",
        headers: jsonHeaders,
        body: JSON.stringify({ role_id: roleId }),
      },
      (value): value is WorldRoleResponse =>
        isWorldRoleFor(value, worldId, roleId, "npc", true),
    );
  },
  async setNpcEnabled(worldId, roleId, enabled) {
    return requestJson(
      `/api/worlds/${worldId}/roles/${roleId}`,
      {
        method: "PATCH",
        headers: jsonHeaders,
        body: JSON.stringify({ enabled }),
      },
      (value): value is WorldRoleResponse =>
        isWorldRoleFor(value, worldId, roleId, "npc", enabled),
    );
  },
  async removeNpc(worldId, roleId) {
    return requestEmpty(`/api/worlds/${worldId}/roles/${roleId}`, {
      method: "DELETE",
    });
  },
  async replaceLocationRules(worldId, roleId, rules) {
    return requestJson(
      `/api/worlds/${worldId}/roles/${roleId}/location-rules`,
      { method: "PUT", headers: jsonHeaders, body: JSON.stringify(rules) },
      (value): value is LocationRuleResponse[] =>
        isRuleListFor(value, worldId, roleId, rules.length),
    );
  },
  async selectLocation(worldId, locationId) {
    return requestJson(
      `/api/worlds/${worldId}/location`,
      {
        method: "POST",
        headers: jsonHeaders,
        body: JSON.stringify({ location_id: locationId }),
      },
      (value): value is GameViewResponse =>
        isGameViewFor(value, worldId, "the_world_map", locationId),
    );
  },
  async getGameView(worldId, sceneId, signal) {
    return requestJson(
      `/api/worlds/${worldId}/game-view?scene_id=${encodeURIComponent(sceneId)}`,
      { headers: { Accept: "application/json" }, signal },
      (value): value is GameViewResponse =>
        isGameViewFor(value, worldId, sceneId),
    );
  },
};

File: backend/src/app/api/worlds.py
"""世界、世界角色、位置规则和游戏视图 HTTP API。"""

from datetime import datetime
from typing import Annotated, Literal, TypeGuard

from fastapi import APIRouter, HTTPException, status
from fastapi.responses import Response
from pydantic import BaseModel, ConfigDict, Field

from app.api.roles import RoleResponse
from app.attribute import AttributeScalar
from app.resources.catalog import AssetNotFoundError, InvalidResourceNameError
from app.world.service import (
    GameViewRecord,
    InvalidLocationError,
    LocationCandidateDraft,
    LocationRuleDraft,
    LocationRuleRecord,
    ProtagonistDraft,
    WorldConflictError,
    WorldNotFoundError,
    WorldRecord,
    WorldRoleRecord,
    WorldService,
)

LocationId = Literal[
    "the_home",
    "the_dungeon",
    "the_mall",
    "the_guild",
    "the_hotel",
    "the_school",
]


class WorldCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    protagonist_name: Annotated[str, Field(min_length=1, max_length=120)]
    protagonist_persona: Annotated[str, Field(min_length=1, max_length=8000)]


class WorldRoleAdd(BaseModel):
    model_config = ConfigDict(extra="forbid")

    role_id: int = Field(gt=0)


class WorldRoleEnabledUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    enabled: bool


class LocationSelection(BaseModel):
    model_config = ConfigDict(extra="forbid")

    location_id: LocationId


class LocationCandidateCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    location_id: LocationId
    weight: int = Field(gt=0)


class LocationRuleCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    weekday_mask: int = Field(ge=1, le=127)
    time_slot: Literal["morning", "midday", "evening", "night"]
    mode: Literal["fixed", "random"]
    priority: int
    enabled: bool = True
    candidates: list[LocationCandidateCreate] = Field(min_length=1)


class LocationRuleResponse(LocationRuleCreate):
    id: int
    world_id: int
    role_id: int


class MapLocationResponse(BaseModel):
    scene_id: LocationId
    display_name: str
    order: int
    anchor_x: float = Field(ge=0, le=1)
    anchor_y: float = Field(ge=0, le=1)


class EffectiveAttributesResponse(BaseModel):
    world_id: int
    role_id: int
    values: dict[str, AttributeScalar]
    updated_at: datetime


class WorldRoleResponse(BaseModel):
    world_id: int
    role_id: int
    name: str
    kind: Literal["player", "npc"]
    enabled: bool
    effective_attributes: dict[str, AttributeScalar]
    portrait_url: str
    version: int
    updated_at: datetime


class WorldResponse(BaseModel):
    id: int
    display_name: str
    active_branch_id: int
    day: int
    weekday: str
    time_slot: str
    player_role: RoleResponse
    npc_count: int
    last_played_at: datetime


class GameViewResponse(BaseModel):
    world_id: int
    scene_id: str
    day: int
    weekday: str
    time_slot: str
    background_url: str
    fallback_background_url: str
    player_marker_url: str
    player_location_id: str
    locations: list[MapLocationResponse]
    visible_roles: list[WorldRoleResponse]


def _world_response(record: WorldRecord) -> WorldResponse:
    return WorldResponse(
        id=record.id,
        display_name=record.display_name,
        active_branch_id=record.active_branch_id,
        day=record.day,
        weekday=record.weekday,
        time_slot=record.time_slot,
        player_role=RoleResponse(
            id=record.player_role.id,
            name=record.player_role.name,
            persona=record.player_role.persona,
            system_prompt=record.player_role.system_prompt,
            world_book=record.player_role.world_book,
            attributes=list(record.player_role.attributes),
            effective_base_values={
                item.key: item.base_value for item in record.player_role.attributes
            },
            portrait_url=record.player_role.portrait_url,
            referenced_world_ids=list(record.player_role.referenced_world_ids),
            version=record.player_role.version,
            created_at=record.player_role.created_at,
            updated_at=record.player_role.updated_at,
        ),
        npc_count=record.npc_count,
        last_played_at=record.last_played_at,
    )


def _world_role_response(record: WorldRoleRecord) -> WorldRoleResponse:
    return WorldRoleResponse(
        world_id=record.world_id,
        role_id=record.role_id,
        name=record.name,
        kind=record.kind,
        enabled=record.enabled,
        effective_attributes=record.effective_attributes,
        portrait_url=record.portrait_url,
        version=record.version,
        updated_at=record.updated_at,
    )


def _location_rule_response(record: LocationRuleRecord) -> LocationRuleResponse:
    if not _is_time_slot(record.time_slot):
        raise RuntimeError("持久化位置规则时间段无效")
    return LocationRuleResponse(
        id=record.id,
        world_id=record.world_id,
        role_id=record.role_id,
        weekday_mask=record.weekday_mask,
        time_slot=record.time_slot,
        mode=record.mode,
        priority=record.priority,
        enabled=record.enabled,
        candidates=[
            LocationCandidateCreate(
                location_id=_required_location_id(item.location_id), weight=item.weight
            )
            for item in record.candidates
        ],
    )


def _game_view_response(record: GameViewRecord) -> GameViewResponse:
    return GameViewResponse(
        world_id=record.world_id,
        scene_id=record.scene_id,
        day=record.day,
        weekday=record.weekday,
        time_slot=record.time_slot,
        background_url=record.background_url,
        fallback_background_url=record.fallback_background_url,
        player_marker_url=record.player_marker_url,
        player_location_id=record.player_location_id,
        locations=[
            MapLocationResponse(
                scene_id=_required_location_id(item.scene_id),
                display_name=item.display_name,
                order=item.order,
                anchor_x=item.anchor_x,
                anchor_y=item.anchor_y,
            )
            for item in record.locations
        ],
        visible_roles=[_world_role_response(item) for item in record.visible_roles],
    )


def _is_time_slot(value: str) -> TypeGuard[Literal["morning", "midday", "evening", "night"]]:
    return value in {"morning", "midday", "evening", "night"}


def _is_location_id(value: str) -> TypeGuard[LocationId]:
    return value in {
        "the_home",
        "the_dungeon",
        "the_mall",
        "the_guild",
        "the_hotel",
        "the_school",
    }


def _required_location_id(value: str) -> LocationId:
    if not _is_location_id(value):
        raise RuntimeError("持久化地图地点无效")
    return value


def _not_found(error: WorldNotFoundError) -> HTTPException:
    return HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(error))


def _conflict(error: WorldConflictError) -> HTTPException:
    return HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(error))


def _invalid_location(error: InvalidLocationError) -> HTTPException:
    return HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_CONTENT, detail=str(error))


def create_world_router(service: WorldService) -> APIRouter:
    """路由只负责严格 DTO、Service 调用和业务错误到 HTTP 的转换。"""

    router = APIRouter(prefix="/api")

    def list_worlds() -> list[WorldResponse]:
        return [_world_response(item) for item in service.list_worlds()]

    def create_world(payload: WorldCreate) -> WorldResponse:
        try:
            return _world_response(
                service.create_world(
                    ProtagonistDraft(
                        name=payload.protagonist_name,
                        persona=payload.protagonist_persona,
                    )
                )
            )
        except AssetNotFoundError, InvalidResourceNameError:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="主角缺少有效的同名默认立绘",
            ) from None
        except WorldConflictError as error:
            raise _conflict(error) from None

    def delete_world(world_id: int) -> Response:
        try:
            service.delete_world(world_id)
        except WorldNotFoundError as error:
            raise _not_found(error) from None
        except WorldConflictError as error:
            raise _conflict(error) from None
        return Response(status_code=status.HTTP_204_NO_CONTENT)

    def list_world_roles(world_id: int) -> list[WorldRoleResponse]:
        try:
            return [_world_role_response(item) for item in service.list_world_roles(world_id)]
        except WorldNotFoundError as error:
            raise _not_found(error) from None

    def add_world_role(world_id: int, payload: WorldRoleAdd) -> WorldRoleResponse:
        try:
            return _world_role_response(service.add_npc(world_id, payload.role_id))
        except AssetNotFoundError, InvalidResourceNameError:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="NPC 缺少有效的同名默认立绘",
            ) from None
        except WorldNotFoundError as error:
            raise _not_found(error) from None
        except WorldConflictError as error:
            raise _conflict(error) from None

    def update_world_role(
        world_id: int,
        role_id: int,
        payload: WorldRoleEnabledUpdate,
    ) -> WorldRoleResponse:
        try:
            return _world_role_response(
                service.update_npc_enabled(world_id, role_id, payload.enabled)
            )
        except WorldNotFoundError as error:
            raise _not_found(error) from None
        except WorldConflictError as error:
            raise _conflict(error) from None

    def remove_world_role(world_id: int, role_id: int) -> Response:
        try:
            service.remove_npc(world_id, role_id)
        except WorldNotFoundError as error:
            raise _not_found(error) from None
        except WorldConflictError as error:
            raise _conflict(error) from None
        return Response(status_code=status.HTTP_204_NO_CONTENT)

    def replace_location_rules(
        world_id: int,
        role_id: int,
        payload: list[LocationRuleCreate],
    ) -> list[LocationRuleResponse]:
        try:
            records = service.replace_location_rules(
                world_id,
                role_id,
                tuple(
                    LocationRuleDraft(
                        weekday_mask=item.weekday_mask,
                        time_slot=item.time_slot,
                        mode=item.mode,
                        priority=item.priority,
                        enabled=item.enabled,
                        candidates=tuple(
                            LocationCandidateDraft(
                                location_id=candidate.location_id,
                                weight=candidate.weight,
                            )
                            for candidate in item.candidates
                        ),
                    )
                    for item in payload
                ),
            )
            return [_location_rule_response(item) for item in records]
        except WorldNotFoundError as error:
            raise _not_found(error) from None
        except WorldConflictError as error:
            raise _conflict(error) from None
        except InvalidLocationError as error:
            raise _invalid_location(error) from None

    def read_effective_attributes(world_id: int, role_id: int) -> EffectiveAttributesResponse:
        try:
            record = service.get_world_role(world_id, role_id)
        except WorldNotFoundError as error:
            raise _not_found(error) from None
        return EffectiveAttributesResponse(
            world_id=record.world_id,
            role_id=record.role_id,
            values=record.effective_attributes,
            updated_at=record.updated_at,
        )

    def select_location(world_id: int, payload: LocationSelection) -> GameViewResponse:
        try:
            return _game_view_response(service.select_location(world_id, payload.location_id))
        except WorldNotFoundError as error:
            raise _not_found(error) from None
        except WorldConflictError as error:
            raise _conflict(error) from None
        except InvalidLocationError as error:
            raise _invalid_location(error) from None

    def get_game_view(world_id: int, scene_id: str) -> GameViewResponse:
        try:
            return _game_view_response(service.get_game_view(world_id, scene_id))
        except WorldNotFoundError as error:
            raise _not_found(error) from None
        except WorldConflictError as error:
            raise _conflict(error) from None
        except InvalidLocationError as error:
            raise _invalid_location(error) from None

    router.add_api_route(
        "/worlds", list_worlds, methods=["GET"], response_model=list[WorldResponse]
    )
    router.add_api_route(
        "/worlds", create_world, methods=["POST"], response_model=WorldResponse, status_code=201
    )
    router.add_api_route("/worlds/{world_id}", delete_world, methods=["DELETE"], status_code=204)
    router.add_api_route(
        "/worlds/{world_id}/roles",
        list_world_roles,
        methods=["GET"],
        response_model=list[WorldRoleResponse],
    )
    router.add_api_route(
        "/worlds/{world_id}/roles",
        add_world_role,
        methods=["POST"],
        response_model=WorldRoleResponse,
        status_code=201,
    )
    router.add_api_route(
        "/worlds/{world_id}/roles/{role_id}",
        update_world_role,
        methods=["PATCH"],
        response_model=WorldRoleResponse,
    )
    router.add_api_route(
        "/worlds/{world_id}/roles/{role_id}",
        remove_world_role,
        methods=["DELETE"],
        status_code=204,
    )
    router.add_api_route(
        "/worlds/{world_id}/roles/{role_id}/location-rules",
        replace_location_rules,
        methods=["PUT"],
        response_model=list[LocationRuleResponse],
    )
    router.add_api_route(
        "/worlds/{world_id}/roles/{role_id}/effective-attributes",
        read_effective_attributes,
        methods=["GET"],
        response_model=EffectiveAttributesResponse,
    )
    router.add_api_route(
        "/worlds/{world_id}/location",
        select_location,
        methods=["POST"],
        response_model=GameViewResponse,
    )
    router.add_api_route(
        "/worlds/{world_id}/game-view",
        get_game_view,
        methods=["GET"],
        response_model=GameViewResponse,
    )
    return router
